### ISSUE TITLE
AES Encrypt and decrypt from Hex

### ISSUE THREAD

**Issue by andrepcg** [brix/crypto-js#143]
What is wrong with the following piece of code?

```javascript
var key = '123';
// Encrypt
var ciphertext = CryptoJS.AES.encrypt('my message', key);
var cypherString = ciphertext.toString(CryptoJS.format.Hex)

// Decrypt
var bytes  = CryptoJS.AES.decrypt(cypherString, key, {format: CryptoJS.format.Hex });
var plaintext = bytes.toString();
```

`plaintext` is always an empty string when it should output 'my message'

**Comment by ifrias-fuze** [comment]
I found that you can actually import the enc object and use that to stringify/parse strings.

I am not sure about hex, but to decrypt a Utf8 formatted string I ended up doing: 

const plainText = enc.Utf8.stringify(cypherString, key);

Note: This would only work if cypherString = cipherText.toString() -> because by default the encrypt use Utf8 encoding.

**Comment by qiudaomao** [comment]
@andrepcg 

try below, the input value should be an object contains ciphertext
// Decrypt
var bytes  = CryptoJS.AES.decrypt(**{
    ciphertext: CryptoJS.enc.Hex.parse(cypherString)
}**, key, {format: CryptoJS.format.Hex });
var plaintext = bytes.toString();

**Comment by ohager** [comment]
@qiudaomao I cannot make this work... any thoughts?

```        const key = '123';
        const ciphertext = CryptoJS.AES.encrypt('my message', key);
        const cypherString = ciphertext.toString(CryptoJS.format.Hex);

        const bytes = CryptoJS.AES.decrypt({
            ciphertext: CryptoJS.enc.Hex.parse(cypherString)
        }, key, {format: CryptoJS.format.Hex});
        const plaintext = bytes.toString();
```

`plaintext` remains empty

**Comment by qiudaomao** [comment]
> @qiudaomao I cannot make this work... any thoughts?
> 
> ```
>         const ciphertext = CryptoJS.AES.encrypt('my message', key);
>         const cypherString = ciphertext.toString(CryptoJS.format.Hex);
> 
>         const bytes = CryptoJS.AES.decrypt({
>             ciphertext: CryptoJS.enc.Hex.parse(cypherString)
>         }, key, {format: CryptoJS.format.Hex});
>         const plaintext = bytes.toString();
> ```
> `plaintext` remains empty

```javascript
//here returns an encrypted object contains `ciphertext`
//js is really stupid, you never known method input/output types without read the code...
const cipherResult = CryptoJS.AES.encrypt('my message', key);
const cipherString = CryptoJS.enc.Hex.stringify(cipherResult.ciphertext);

const bytes = CryptoJS.AES.decrypt({
    ciphertext: CryptoJS.enc.Hex.parse(cypherString)
},  key,  {format: CryptoJS.format.Hex});
const plaintext = bytes.toString();
```

**Comment by TOPSinfo** [comment]
var data = [{id: 1}, {id: 2}]
    // Encrypt
    var ciphertext = CryptoJS.AES.encrypt(JSON.stringify(data), key);
    const cypherString = ciphertext.toString(CryptoJS.format.Hex);
    console.log("cipherText here is",cypherString);
    
    var bytes  = CryptoJS.AES.decrypt(ciphertext.toString(), key);
    var decryptedData = JSON.parse(bytes.toString(CryptoJS.enc.Utf8));

**Comment by Barbery** [comment]
I meet the problem too, anyone know how to solve it?

**Comment by 92sajid110** [comment]
no solution yet

**Comment by belaarany** [comment]
Still no solution...

**Comment by yxw007** [comment]
```
var key = CryptoJS.enc.Utf8.parse("4ozcn613nu8irs5x");
        var plaintText = '123'; 
        //encrypt
        var encryptedData = CryptoJS.AES.encrypt(plaintText, key, {
          mode: CryptoJS.mode.ECB,
          padding: CryptoJS.pad.Pkcs7
        });
        console.log("plaintText：" + plaintText);
        console.log("encryptedData ：" + encryptedData);

        var encryptedDataHexStr = encryptedData.toString(CryptoJS.format.Hex);
        console.log("encryptedDataHex：" + encryptedDataHexStr );
        //-------------------------------------------------------------------------
        //decrypt 
        var encryptedHex = CryptoJS.enc.Hex.parse(encryptedDataHexStr);
        var encryptedBase64 = CryptoJS.enc.Base64.stringify(encryptedHex);
        console.log("encryptedHex  ：" + encryptedHex);
        console.log("encryptedBase64  ：" + encryptedBase64 );

        var decryptedData = CryptoJS.AES.decrypt(encryptedBase64, key, {
          mode: CryptoJS.mode.ECB,
          padding: CryptoJS.pad.Pkcs7,
        });

        let decryptText = decryptedData.toString(CryptoJS.enc.Utf8);
        console.log("decryptText ：" + decryptText );
```
        
   This is my solution, share it with everyone

**Comment by ghost** [comment]
@aa4790139 bro, do you see that your identifiers are all inconsistent and confusing.

I'm pretty sure encryptedDataHexStr is the same as encryptedDataHex and encryptedHexStr since the latter were never declared. But it's hard to follow like that.

**Comment by yxw007** [comment]
> @aa4790139 bro, do you see that your identifiers are all inconsistent and confusing.
> 
> I'm pretty sure encryptedDataHexStr is the same as encryptedDataHex and encryptedHexStr since the latter were never declared. But it's hard to follow like that.

sorry decrypt edit error， already corrected

**Comment by 92Sam** [comment]
This work for me, using enc and not only UTF8

```
const { AES, enc } = require('crypto-js')
const { decrypt, encrypt } = AES


const message = "Hi my friend"
const key= "key123456"
const aesDecrypt = decrypt(message, key).toString(enc.Utf8)
console.log(aesDecrypt) // Hi my friend
```

**Comment by HereOrCode** [comment]
> ```
> var key = CryptoJS.enc.Utf8.parse("4ozcn613nu8irs5x");
>         var plaintText = '123'; 
>         //encrypt
>         var encryptedData = CryptoJS.AES.encrypt(plaintText, key, {
>           mode: CryptoJS.mode.ECB,
>           padding: CryptoJS.pad.Pkcs7
>         });
>         console.log("plaintText：" + plaintText);
>         console.log("encryptedData ：" + encryptedData);
> 
>         var encryptedDataHexStr = encryptedData.toString(CryptoJS.format.Hex);
>         console.log("encryptedDataHex：" + encryptedDataHexStr );
>         //-------------------------------------------------------------------------
>         //decrypt 
>         var encryptedHex = CryptoJS.enc.Hex.parse(encryptedDataHexStr);
>         var encryptedBase64 = CryptoJS.enc.Base64.stringify(encryptedHex);
>         console.log("encryptedHex  ：" + encryptedHex);
>         console.log("encryptedBase64  ：" + encryptedBase64 );
> 
>         var decryptedData = CryptoJS.AES.decrypt(encryptedBase64, key, {
>           mode: CryptoJS.mode.ECB,
>           padding: CryptoJS.pad.Pkcs7,
>         });
> 
>         let decryptText = decryptedData.toString(CryptoJS.enc.Utf8);
>         console.log("decryptText ：" + decryptText );
> ```
> 
> This is my solution, share it with everyone

Working for me, thanks

**Comment by imarthurg** [comment]
All the examples above didn't work for me, but this one works:

`const SECRET = 'I am batman'

function enc(plainText){
    var b64 = CryptoJS.AES.encrypt(plainText, SECRET).toString();
    var e64 = CryptoJS.enc.Base64.parse(b64);
    var eHex = e64.toString(CryptoJS.enc.Hex);
    return eHex;
}

function dec(cipherText){
   var reb64 = CryptoJS.enc.Hex.parse(cipherText);
   var bytes = reb64.toString(CryptoJS.enc.Base64);
   var decrypt = CryptoJS.AES.decrypt(bytes, SECRET);
   var plain = decrypt.toString(CryptoJS.enc.Utf8);
   return plain;
}`
Reference: https://itecnote.com/tecnote/javascript-cipher-a-string-using-crypto-js-with-hex-encoding-to-make-it-url-friendly/

In my case, I need a hex from an encrypted plain text, and the plain text when decrypting that generated hex. eg.:

```
const plainText = "Hello World";

console.log(enc(plainText)); 
// > "<the_generated_hex>"

console.log(dec("<the_generated_hex>"));
// > "Hello World"

```
Hope this is helpful :)

**Comment by R4tch3t** [comment]
> ```
> var key = CryptoJS.enc.Utf8.parse("4ozcn613nu8irs5x");
>         var plaintText = '123'; 
>         //encrypt
>         var encryptedData = CryptoJS.AES.encrypt(plaintText, key, {
>           mode: CryptoJS.mode.ECB,
>           padding: CryptoJS.pad.Pkcs7
>         });
>         console.log("plaintText：" + plaintText);
>         console.log("encryptedData ：" + encryptedData);
> 
>         var encryptedDataHexStr = encryptedData.toString(CryptoJS.format.Hex);
>         console.log("encryptedDataHex：" + encryptedDataHexStr );
>         //-------------------------------------------------------------------------
>         //decrypt 
>         var encryptedHex = CryptoJS.enc.Hex.parse(encryptedDataHexStr);
>         var encryptedBase64 = CryptoJS.enc.Base64.stringify(encryptedHex);
>         console.log("encryptedHex  ：" + encryptedHex);
>         console.log("encryptedBase64  ：" + encryptedBase64 );
> 
>         var decryptedData = CryptoJS.AES.decrypt(encryptedBase64, key, {
>           mode: CryptoJS.mode.ECB,
>           padding: CryptoJS.pad.Pkcs7,
>         });
> 
>         let decryptText = decryptedData.toString(CryptoJS.enc.Utf8);
>         console.log("decryptText ：" + decryptText );
> ```
> 
> This is my solution, share it with everyone

Greats works for me, ok if changes utf16 works for parse a large keys
